src/shims/posix/linux/foreign_items.rs

```diff
@@ -192,3 +201,61 @@ fn getrandom<'tcx>(
     this.write_scalar(Scalar::from_machine_usize(len, this), dest)?;
     Ok(())
 }
+
+fn futex<'tcx>(
+    this: &mut MiriEvalContext<'_, 'tcx>,
+    args: &[OpTy<'tcx, Tag>],
+    dest: PlaceTy<'tcx, Tag>,
+) -> InterpResult<'tcx> {
+    if args.len() < 4 {
+        throw_ub_format!("incorrect number of arguments for futex syscall: got {}, expected at least 4", args.len());
+    }
+    let addr = this.read_scalar(args[1])?.check_init()?;
+    let op = this.read_scalar(args[2])?.to_i32()?;
+    let val = this.read_scalar(args[3])?.to_i32()?;
+
+    this.memory.check_ptr_access(addr, Size::from_bytes(4), Align::from_bytes(4).unwrap())?;
+
+    let addr = addr.assert_ptr();
+
+    let thread = this.get_active_thread();
+
+    let futex_private = this.eval_libc_i32("FUTEX_PRIVATE_FLAG")?;
+    let futex_wait = this.eval_libc_i32("FUTEX_WAIT")?;
+    let futex_wake = this.eval_libc_i32("FUTEX_WAKE")?;
+
+    match op & !futex_private {
+        op if op == futex_wait => {
+            if args.len() < 5 {
+                throw_ub_format!("incorrect number of arguments for FUTEX_WAIT syscall: got {}, expected at least 5", args.len());
+            }
+            let timeout = this.read_scalar(args[4])?.check_init()?;
```

Comment:
So this is a vararg situation? Please check `fnctl` for how we handle those... everything should call `check_arg_count` eventually.

Reply:
I don't think check_arg_count would work here. The futex() syscall takes six arguments, but depending on the second argument it might ignore the last few ones. Some callers always pass all six, and some callers pass only the kernel actually reads. So an operation like FUTEX_WAKE might be called with 3, 4, 5 or 6 arguments (excluding the syscall number itself).

Reply:
Oh so it is legal to pass more arguments than expected? That's weird, I would think this is a problem wrt. calling conventions...

Please at least check that it is no more than 6, and add a comment explaining why there is "wiggle room" here. I don't think we have any other shim that behaves like this.

Reply:
Pretty sure that's legal for all vararg functions. `printf("%d %d", 1, 2, 3, 4, 5);` should work fine. For SYS_futex specifically: Both std and parking_lot_core pass the minimal number of arguments. But I've seen code that wraps the entire SYS_futex call first with all six parameters, and uses that for all futex operations.